### ISSUE TITLE
DynamicObjectLeak - questions

### ISSUE THREAD

**Issue by clement-brodu** [Riverside-Software/sonar-openedge#1121]
Hello, 

We have many “false positives” with this rule.

Sometimes our DELETE OBJECTs are not in the same block.
They are often found in an IF statement, or in other sub-blocks (see 2 examples above), but the DELETE OBJECT is at the end of the method.

We fully understand that it is not easy to detect these cases, but do you think it is possible to improve this rule to avoid these cases which are false positives for us.

Of course, for some of our cases, we could refactor our code to avoid getting the error, but this is not always easy.

The rule could be applied to the variable definition block (the method/function/procedure).

What do you think ?

```abl
DEFINE TEMP-TABLE ttDemo NO-UNDO
            FIELD Order AS INTEGER  
            FIELD Name AS CHARACTER   
            INDEX idx1 IS PRIMARY UNIQUE Name
            INDEX idx2 Order Name.


    METHOD PUBLIC VOID Case1(INPUT pRefresh AS LOGICAL):
        DEFINE VARIABLE vhQuery AS HANDLE NO-UNDO.
        DEFINE VARIABLE vhBuffer AS HANDLE NO-UNDO.
        
        IF pRefresh THEN
        DO:
            CREATE QUERY vhQuery. // Sonar: false positive Memory allocated 
            CREATE BUFFER vhBuffer FOR TABLE TEMP-TABLE ttDemo:HANDLE. // Sonar: false positiveMemory allocated
            vhQuery:SET-BUFFERS (vhBuffer).

            vhQuery:QUERY-PREPARE ("FOR EACH ttDemo WHERE ttDemo.order > 5").
            vhQuery:QUERY-OPEN().
            vhQuery:GET-FIRST(NO-LOCK).
            DO WHILE NOT vhQuery:QUERY-OFF-END:    
                // Do stuff 
                
                vhQuery:GET-NEXT(NO-LOCK).
            END.
            vhQuery:QUERY-CLOSE ().
        END.
        
        FINALLY:
            IF VALID-HANDLE(vhQuery) THEN DELETE OBJECT vhQuery.
            IF VALID-HANDLE(vhBuffer) THEN DELETE OBJECT vhBuffer. 
        END FINALLY.
    END METHOD.
    

    METHOD PUBLIC VOID Case2(  ):
        DEFINE VARIABLE vii AS INTEGER NO-UNDO.
        DEFINE VARIABLE vhQuery AS HANDLE NO-UNDO.
        DEFINE VARIABLE vhBuffer AS HANDLE NO-UNDO.

        
        DO vii = 1 TO 10:
            IF NOT VALID-HANDLE(vhQuery) THEN
            DO:
                CREATE QUERY vhQuery. // Sonar: false positive Memory allocated 
                CREATE BUFFER vhBuffer FOR TABLE TEMP-TABLE ttDemo:HANDLE. // Sonar: false positiveMemory allocated
                vhQuery:SET-BUFFERS (vhBuffer).
            END.
            
            vhQuery:QUERY-PREPARE (SUBSTITUTE("FOR EACH ttDemo WHERE ttDemo.order = &1", vii)).
            vhQuery:QUERY-OPEN().
            vhQuery:GET-FIRST(NO-LOCK).
            DO WHILE NOT vhQuery:QUERY-OFF-END:    
                // Do stuff 
                
                vhQuery:GET-NEXT(NO-LOCK).
            END.
            vhQuery:QUERY-CLOSE ().
        END.
        
        FINALLY:
            IF VALID-HANDLE(vhQuery) THEN DELETE OBJECT vhQuery.
            IF VALID-HANDLE(vhBuffer) THEN DELETE OBJECT vhBuffer. 
        END FINALLY.

    END METHOD.

    METHOD PUBLIC VOID Case3 ( INPUT pNomTableEntity AS CHARACTER, INPUT pTempTableHandle AS HANDLE, INPUT pRowid AS ROWID ):
        
        DEFINE VARIABLE vhBuffer       AS HANDLE    NO-UNDO.
        DEFINE VARIABLE vNomBuffer     AS CHARACTER NO-UNDO.
        
        vNomBuffer = "buffer":U. 
        IF pTempTableHandle = ? THEN 
            CREATE BUFFER vhBuffer FOR TABLE pNomTableEntity BUFFER-NAME vNomBuffer.
        ELSE 
            CREATE BUFFER vhBuffer FOR TABLE pTempTableHandle BUFFER-NAME vNomBuffer.
        
        vhBuffer:FIND-BY-ROWID(pRowid, NO-LOCK) NO-ERROR. 
        
        /* Do Stuff */
        
        FINALLY:
            IF VALID-OBJECT(vhBuffer) THEN DELETE OBJECT vhBuffer NO-ERROR.    
        END FINALLY.

    END METHOD.
```

**Comment by gquerret** [comment]
Yes, this one has already been fixed. Next release end of week.

**Comment by Ton-Pe** [comment]
Great! I was just coming to report this same issue :) 
Anyway, we do have a helper class that we use to clean up dynamic objects, class instances etc. and understandably the rule does not understand this. 

For example:
`CREATE QUERY hQuery.`
`GarbageCollectorHelper:DeleteObject(hQuery).`

But I'll create an improvement idea for this.

**Comment by Ton-Pe** [comment]
@gquerret ,
I noticed that we have lot's of old code where dynamic objects/buffers are deleted using `DELETE WIDGET` command instead of `DELETE OBJECT`. This rule does not recognize these so it gives a false positive.


BLOCK-LEVEL ON ERROR UNDO, THROW.

DEFINE VARIABLE xObj AS HANDLE NO-UNDO.
DEFINE VARIABLE xWid AS HANDLE NO-UNDO.

CREATE BUFFER xObj FOR TABLE "Item" NO-ERROR.    
CREATE BUFFER xWid FOR TABLE "Customer" NO-ERROR. /* this line is flagged*/

FINALLY:
    IF VALID-HANDLE (xObj) THEN DELETE OBJECT xObj.
    IF VALID-HANDLE (xWid) THEN DELETE WIDGET xWid. /* this is not understood as deletion */
END.
 
 
 Of course, in this case xWid is not a real widget. What to you think, could you improve the rule still?
